File: androidApp/src/main/kotlin/com/slowbuild/storyverse/storyverse/ui/search/SearchViewModel.kt
package com.slowbuild.storyverse.storyverse.ui.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.slowbuild.storyverse.core.result.AppResult
import com.slowbuild.storyverse.domain.model.Story
import com.slowbuild.storyverse.domain.source.StoryFilter
import com.slowbuild.storyverse.domain.source.StorySourceRegistry
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SearchUiState(
    val query: String = "",
    val isLoading: Boolean = false,
    val searchResults: List<Story> = emptyList(),
    val availableCategories: List<String> = listOf("Tiên Hiệp", "Kiếm Hiệp", "Huyền Huyễn", "Đô Thị", "Khoa Huyễn", "Lịch Sử", "Võng Du"),
    val selectedCategory: String? = null,
    val hasSearched: Boolean = false,
    val errorMessage: String? = null
)

class SearchViewModel(
    private val storySourceRegistry: StorySourceRegistry
) : ViewModel() {

    private val _uiState = MutableStateFlow(SearchUiState())
    val uiState: StateFlow<SearchUiState> = _uiState.asStateFlow()

    private var searchJob: Job? = null

    fun onQueryChange(newQuery: String) {
        _uiState.update { it.copy(query = newQuery) }
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            delay(300) // Debounce
            executeSearch(query = newQuery, category = _uiState.value.selectedCategory)
        }
    }

    fun onCategorySelect(category: String?) {
        val nextCategory = if (_uiState.value.selectedCategory == category) null else category
        _uiState.update { it.copy(selectedCategory = nextCategory) }
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            executeSearch(query = _uiState.value.query, category = nextCategory)
        }
    }

    private suspend fun executeSearch(query: String, category: String?) {
        if (query.isBlank() && category == null) {
            _uiState.update {
                it.copy(
                    isLoading = false,
                    searchResults = emptyList(),
                    hasSearched = false,
                    errorMessage = null
                )
            }
            return
        }

        _uiState.update { it.copy(isLoading = true, hasSearched = true, errorMessage = null) }
        val source = storySourceRegistry.getDefaultSource() ?: return

        val filter = if (category != null) StoryFilter(category = category) else null
        when (val result = source.search(query = query, page = 1, filter = filter)) {
            is AppResult.Success -> {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        searchResults = result.data.stories,
                        errorMessage = null
                    )
                }
            }
            is AppResult.Error -> {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        errorMessage = result.error.message
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/ViewModels/SearchViewModel.swift
import Foundation
import SharedLogic

@MainActor
public final class SearchViewModel: ObservableObject {
    @Published public var query: String = ""
    @Published public var selectedCategory: String? = nil
    @Published public private(set) var availableCategories: [String] = [
        "Tiên Hiệp", "Kiếm Hiệp", "Huyền Huyễn", "Đô Thị", "Khoa Huyễn", "Lịch Sử", "Võng Du"
    ]
    @Published public private(set) var results: [Story] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var isLoadingMore: Bool = false
    @Published public private(set) var hasNextPage: Bool = false
    @Published public private(set) var currentPage: Int = 1
    @Published public private(set) var errorMessage: String? = nil

    private let storySourceRegistry: StorySourceRegistry
    private var searchTask: Task<Void, Never>?

    public init(storySourceRegistry: StorySourceRegistry = KoinHelper().storySourceRegistry) {
        self.storySourceRegistry = storySourceRegistry
    }

    public func onQueryChange(_ newQuery: String) {
        query = newQuery
        searchTask?.cancel()
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000) // 300ms debounce
            if !Task.isCancelled {
                performSearch(query: newQuery, category: selectedCategory, page: 1)
            }
        }
    }

    public func onCategorySelect(_ category: String?) {
        let nextCategory = (selectedCategory == category) ? nil : category
        selectedCategory = nextCategory
        searchTask?.cancel()
        searchTask = Task {
            performSearch(query: query, category: nextCategory, page: 1)
        }
    }

    public func loadNextPage() {
        guard !isLoading && !isLoadingMore && hasNextPage else { return }
        guard let source = storySourceRegistry.getDefaultSource() else { return }

        let nextPage = currentPage + 1
        isLoadingMore = true

        Task {
            do {
                let filter: StoryFilter? = (selectedCategory != nil) ? StoryFilter(category: selectedCategory, status: nil, sort: .popular, author: nil) : nil
                let searchResult = try await source.search(query: query, page: Int32(nextPage), filter: filter)
                if let page = searchResult.getOrNull() as? StoryPage {
                    self.results.append(contentsOf: page.stories)
                    self.currentPage = nextPage
                    self.hasNextPage = page.hasNextPage
                }
                self.isLoadingMore = false
            } catch {
                self.isLoadingMore = false
            }
        }
    }

    public func performSearch(query: String? = nil, category: String? = nil, page: Int = 1) {
        let q = query ?? self.query
        let cat = category ?? self.selectedCategory

        if q.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && cat == nil {
            results = []
            isLoading = false
            isLoadingMore = false
            hasNextPage = false
            currentPage = 1
            errorMessage = nil
            return
        }

        guard let source = storySourceRegistry.getDefaultSource() else {
            self.errorMessage = "No active story source."
            return
        }

        isLoading = true
        errorMessage = nil
        currentPage = page

        Task {
            do {
                let filter: StoryFilter? = (cat != nil) ? StoryFilter(category: cat, status: nil, sort: .popular, author: nil) : nil
                let searchResult = try await source.search(query: q, page: Int32(page), filter: filter)
                if let page = searchResult.getOrNull() as? StoryPage {
                    self.results = page.stories
                    self.hasNextPage = page.hasNextPage
                } else if let err = searchResult.errorOrNull() {
                    self.errorMessage = err.message
                }
                self.isLoading = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }
}
